### ISSUE TITLE
INPC010 should not warn here

### ISSUE THREAD

**Issue by JohanLarsson** [DotNetAnalyzers/PropertyChangedAnalyzers#102]
```cs
namespace RoslynSandbox
{
    using System;
    using System.ComponentModel;
    using System.Runtime.CompilerServices;

    public interface IC
    {
        INotifyPropertyChanged Source { get; set; }
    }

    public sealed class C<T> : INotifyPropertyChanged, IDisposable, IC
         where T : class, INotifyPropertyChanged
    {
        private readonly PropertyChangedEventHandler onTrackedPropertyChanged;
        private readonly object gate = new object();

        private T source;
        private bool disposed;

        public event PropertyChangedEventHandler PropertyChanged;

        public T Source
        {
            get => this.source;

            set
            {
                if (this.disposed)
                {
                    return;
                }

                lock (this.gate)
                {
                    if (this.disposed ||
                        ReferenceEquals(value, this.source))
                    {
                        return;
                    }

                    if (this.source != null)
                    {
                        this.source.PropertyChanged -= this.onTrackedPropertyChanged;
                    }

                    if (value != null)
                    {
                        value.PropertyChanged += this.onTrackedPropertyChanged;
                    }

                    this.source = value;
                    this.OnPropertyChanged();
                }
            }
        }

        INotifyPropertyChanged IC.Source
        {
            get => this.source;
            set => this.Source = (T)value;
        }

        private void OnPropertyChanged([CallerMemberName] string propertyName = null)
        {
            this.PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
        }

        public void Dispose()
        {
            if (this.disposed)
            {
                return;
            }

            INotifyPropertyChanged oldSource;
            lock (this.gate)
            {
                if (this.disposed)
                {
                    return;
                }

                this.disposed = true;
                oldSource = this.source;
                this.source = null;
            }

            if (oldSource != null)
            {
                oldSource.PropertyChanged -= this.onTrackedPropertyChanged;
            }
        }
    }
}
```
